lib/executor/helpers.go

```diff
@@ -84,26 +87,56 @@ func getIterationRunner(
 ) func(context.Context, lib.ActiveVU) bool {
 	return func(ctx context.Context, vu lib.ActiveVU) bool {
 		err := vu.RunOnce()
+		state := vu.GetState()
+		tags := state.CloneTags()
 
 		// TODO: track (non-ramp-down) errors from script iterations as a metric,
 		// and have a default threshold that will abort the script when the error
 		// rate exceeds a certain percentage
 
 		select {
 		case <-ctx.Done():
-			// Don't log errors or emit iterations metrics from cancelled iterations
+			// TODO: Do this elsewhere? It might be racy with
+			// Engine.processMetrics() as the channel could be closed at this point.
+			// TODO: Use enum?
+			// TODO: How to distinguish interruptions from signal (^C)?
+			// tags["cause"] = "duration"
+			// state.Samples <- stats.Sample{
```

Comment:
This is where we would handle the 3rd scenario of executors interrupting the iteration, and while it worked in manual tests and `TestMetricsEmission`, it would cause other tests to hang (presumably because they were setting a `nil` channel), and I'm not sure if it wouldn't be racy, so I commented it out for now.

Is this the right place/approach, or would we have to receive two contexts here or something more complicated?

Reply:
Digging through the context chains we're using, this shouldn't be racy since the `Samples` channel is closed only when the `globalCtx` is done, which happens after all metrics processing is finished, so it should be usable even if this VU "run context" is done. Yet some of our tests use the same context for both, e.g.:

https://github.com/loadimpact/k6/blob/420dd4161c280bdf0398af5e58f73c939057695a/api/v1/status_routes_test.go#L108

This explains why it works in real-world tests but not in the test suite. I'll see if I can fix the tests then.